Data/ByteString/Streaming.hs

```diff
@@ -1195,55 +1194,39 @@ split w = loop
 --
 -- It is a special case of 'groupBy', which allows the programmer to
 -- supply their own equality test.
-
 group :: Monad m => ByteString m r -> Stream (ByteString m) m r
 group = go
   where
-  go (Empty r)         = Return r
-  go (Go m)            = Effect $ liftM go m
-  go (Chunk c cs)
-    | S.length c == 1  = Step $ to [c] (S.unsafeHead c) cs
-    | otherwise        = Step $ to [S.unsafeTake 1 c] (S.unsafeHead c)
-                                     (Chunk (S.unsafeTail c) cs)
-
-  to acc !_ (Empty r)        = revNonEmptyChunks
-                                     acc
-                                     (Empty (Return r))
-  to acc !w (Chunk c cs) =
-    case findIndexOrEnd (/= w) c of
-      0                    -> revNonEmptyChunks
-                                    acc
-                                    (Empty (go (Chunk c cs)))
-      n | n == S.length c  -> to (S.unsafeTake n c : acc) w cs
-        | otherwise        -> revNonEmptyChunks
-                                    (S.unsafeTake n c : acc)
-                                    (Empty (go (Chunk (S.unsafeDrop n c) cs)))
-
+    go (Empty r)        = Return r
+    go (Go m)           = Effect $ liftM go m
+    go (Chunk c cs)
+      | S.length c == 1 = Step $ to [c] (S.unsafeHead c) cs
+      | otherwise       = Step $ to [S.unsafeTake 1 c] (S.unsafeHead c) (Chunk (S.unsafeTail c) cs)
+
+    to acc !_ (Empty r) = revNonEmptyChunks acc (Empty (Return r))
+    to _ _ (Go _) = error "Impossible case"
```

Comment:
This branch of the pattern match was missing. I'm not exactly comfortable with an `error` here, but I'm unsure what was intended by the original code.

Reply:
It seems you've found a bug.  That case is NOT impossible.  Here's an example that crashes `group`:
```haskell
{-# LANGUAGE OverloadedStrings #-}
module Main (main) where
import qualified Data.ByteString.Streaming as Q
import qualified Data.ByteString.Streaming.Internal as Q
import qualified Streaming as S
import qualified System.IO as IO

main :: IO ()
main = do
    let x = Q.Chunk "f" (Q.Go (pure $ Q.Empty ()))
        y = Q.group x
    S.mapsM_ (Q.hPut IO.stdout) y
```
Thus I get:
```
[1 of 3] Compiling Data.ByteString.Streaming.Internal ( Data/ByteString/Streaming/Internal.hs, interpreted )
[2 of 3] Compiling Data.ByteString.Streaming ( Data/ByteString/Streaming.hs, interpreted )
m[3 of 3] Compiling Main             ( foo.hs, interpreted )
Ok, three modules loaded.
λ> main
*** Exception: Data/ByteString/Streaming.hs:(1209,3)-(1220,78): Non-exhaustive patterns in function to
```

Reply:
I see the same problem in `groupBy`. Neither have _ever_ been correct. These look like they're not much used, and so poorly tested...

Reply:
Here's another sample program:
```haskell
{-# LANGUAGE OverloadedStrings #-}
module Main (main) where
import qualified Data.ByteString.Streaming as Q
import qualified Data.ByteString.Streaming.Char8 as Q8
import qualified Data.ByteString.Streaming.Internal as Q
import qualified Streaming as S
import qualified System.IO as IO

main :: IO ()
main = do
    let x = Q.Chunk "a"
          $ Q.Go
          $ pure
          $ Q.Chunk "a"
          $ Q.Go
          $ pure
          $ Q.Chunk "bbc"
          $ Q.Go
          $ pure
          $ Q.Chunk "c"
          $ Q.Go
          $ pure
          $ Q.Empty ()
        y = Q.group x
    S.mapsM_ Q8.putStrLn y
```
and with the patch below it produces the expected:
```
[1 of 4] Compiling Data.ByteString.Streaming.Internal ( Data/ByteString/Streaming/Internal.hs, interpreted )
[2 of 4] Compiling Data.ByteString.Streaming ( Data/ByteString/Streaming.hs, interpreted )
[3 of 4] Compiling Data.ByteString.Streaming.Char8 ( Data/ByteString/Streaming/Char8.hs, interpreted )
[4 of 4] Compiling Main             ( foo.hs, interpreted )
Ok, four modules loaded.
λ> main
aa
bb
cc
```
Proposed patch:
```haskell
diff --git a/Data/ByteString/Streaming.hs b/Data/ByteString/Streaming.hs
index b696c24..be77b51 100644
--- a/Data/ByteString/Streaming.hs
+++ b/Data/ByteString/Streaming.hs
@@ -1206,17 +1206,13 @@ group = go
     | otherwise        = Step $ to [S.unsafeTake 1 c] (S.unsafeHead c)
                                      (Chunk (S.unsafeTail c) cs)
 
-  to acc !_ (Empty r)        = revNonEmptyChunks
-                                     acc
-                                     (Empty (Return r))
+  to acc !_ (Empty r)    = revNonEmptyChunks acc (Empty (Return r))
+  to acc !w (Go m)       = Go $ to acc w <$> m
   to acc !w (Chunk c cs) =
     case findIndexOrEnd (/= w) c of
-      0                    -> revNonEmptyChunks
-                                    acc
-                                    (Empty (go (Chunk c cs)))
+      0                    -> revNonEmptyChunks acc (Empty (go (Chunk c cs)))
       n | n == S.length c  -> to (S.unsafeTake n c : acc) w cs
-        | otherwise        -> revNonEmptyChunks
-                                    (S.unsafeTake n c : acc)
+        | otherwise        -> revNonEmptyChunks (S.unsafeTake n c : acc)
                                     (Empty (go (Chunk (S.unsafeDrop n c) cs)))
 
 {-# INLINABLE group #-}
```

That is, I think the missing branch should be:
```haskell
    to acc !w (Go m)       = Go $ to acc w <$> m
```

Reply:
It seems easy to trigger the bug when manually using the `Streaming.ByteString` constructors, but what about when this is coming from a string literal or a file? What concrete string-based example would trigger this?

Reply:
Well, any streaming bytestring that has a multi-chunk "skeleton" requiring monadic steps should do.  So if you take a file larger than 32k bytes, this should break trying to determine whether the last byte of the first chunk is matched by the first byte of the next chunk (which requires an IO operation to bring it into memory).

Reply:
I'll produce a test file and add a unit test.

Reply:
If you open the file with a smaller (non-default) I/O buffer size, the test file can be smaller...

Reply:
This makes it fail:
```haskell
groupCrash :: Assertion
groupCrash = do
  res <- runResourceT . S.length_ . SM.mapsM SBS8.length . SBS8.group @(ResourceT IO) $ SBS8.readFile "tests/groupBy.txt"
  res @?= 39925
```

Reply:
Did you meat for the `S.length_` to be `S.sum_`?

Reply:
It now is. When I said "fail" I meant "crash" due to the bad pattern match. It's all fixed now! :)